src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -1894,5 +1882,232 @@ struct S
+                    // (15,18): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match.
+                    //             case { Length: 1 } and [.., 1]:
+                    Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "{ Length: 1 } and [.., 1]").WithLocation(15, 18)
+                    );
+        }
+
+        [Fact]
+        public void Subsumption_04()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        switch (a)
+        {
+            case [1, .., 3]:
+            case [1, 2, 3]:
+                break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics(
+                // (9,18): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match.
+                //             case [1, 2, 3]:
+                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "[1, 2, 3]").WithLocation(9, 18)
+                );
+        }
+
+        [Fact]
+        public void Subsumption_05()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        switch (a)
+        {
+            case [1, 2, 3]:
+            case [1, .., 3]:
+                break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void Subsumption_06()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        switch (a)
+        {
+            case [42]:
+            case [..,42]:
+                break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void Subsumption_07()
+        {
+            var src = @"
+class C
+{
+    void Test(int[] a)
+    {
+        switch (a)
+        {
+            case [>0, ..]:
+            case [.., <=0]:
+            case [var unreachable]:
+                    break;
+        }
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRange(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics(
+                // (10,18): error CS8120: The switch case is unreachable. It has already been handled by a previous case or it is impossible to match.
+                //             case [var unreachable]:
+                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "[var unreachable]").WithLocation(10, 18));
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void Subsumption_Slice_00(
```

Comment:
I took one permutation of `Subsumption_Slice_00` and generated a dump:

I'm not sure why `case [..[..[1,2,3]]]:` involves a `length >= 0` check (`t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueDefault(0: Int32))`). Is that correct?

```
State 0
    1. [case [..[..[1,2,3]]]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueDefault(0: Int32)), t2=DagSliceEvaluation(t0), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueDefault(0: Int32)), t4=DagSliceEvaluation(t2), t5=DagPropertyEvaluation(t4.Length), ?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t1=DagPropertyEvaluation(t0.Length)
    TrueBranch: 1

State 1
    1. [case [..[..[1,2,3]]]:] AND(?DagRelationalTest(t1 >= ConstantValueDefault(0: Int32)), t2=DagSliceEvaluation(t0), t3=DagPropertyEvaluation(t2.Length), ?DagRelationalTest(t3 >= ConstantValueDefault(0: Int32)), t4=DagSliceEvaluation(t2), t5=DagPropertyEvaluation(t4.Length), ?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: ?DagRelationalTest(t1 >= ConstantValueDefault(0: Int32))
    TrueBranch: 2
    FalseBranch: 14

State 2 REMAINING t1:[0..2147483647]
    1. [case [..[..[1,2,3]]]:] AND(t2=DagSliceEvaluation(t0), t3=DagPropertyEvaluation(t2.Length), t4=DagSliceEvaluation(t2), t5=DagPropertyEvaluation(t4.Length), ?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t2=DagSliceEvaluation(t0)
    TrueBranch: 3

State 3 REMAINING t1:[0..2147483647]
    1. [case [..[..[1,2,3]]]:] AND(t3=DagPropertyEvaluation(t2.Length), t4=DagSliceEvaluation(t2), t5=DagPropertyEvaluation(t4.Length), ?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t3=DagPropertyEvaluation(t2.Length)
    TrueBranch: 4

State 4 REMAINING t1:[0..2147483647]
    1. [case [..[..[1,2,3]]]:] AND(t4=DagSliceEvaluation(t2), t5=DagPropertyEvaluation(t4.Length), ?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t4=DagSliceEvaluation(t2)
    TrueBranch: 5

State 5 REMAINING t1:[0..2147483647]
    1. [case [..[..[1,2,3]]]:] AND(t5=DagPropertyEvaluation(t4.Length), ?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t5=DagPropertyEvaluation(t4.Length)
    TrueBranch: 6

State 6 REMAINING t1:[0..2147483647]
    1. [case [..[..[1,2,3]]]:] AND(?DagValueTest(t5 == ConstantValueI32(3: Int32)), t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(?DagRelationalTest(t1 >= ConstantValueI32(3: Int32)), t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), ?DagValueTest(t11 == ConstantValueDefault(0: Int32)), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: ?DagValueTest(t5 == ConstantValueI32(3: Int32))
    TrueBranch: 7
    FalseBranch: 14

State 7 REMAINING t1:[3..3] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] AND(t6=DagIndexerEvaluation(t4[0]), ?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t6=DagIndexerEvaluation(t4[0])
    TrueBranch: 8

State 8 REMAINING t1:[3..3] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] AND(?DagValueTest(t6 == ConstantValueOne(1: Int32)), t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(t9=DagIndexerEvaluation(t0[0]), ?DagValueTest(t9 == ConstantValueOne(1: Int32)), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: ?DagValueTest(t6 == ConstantValueOne(1: Int32))
    TrueBranch: 9
    FalseBranch: 14

State 9 REMAINING t6:[1..1] t1:[3..3] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] AND(t7=DagIndexerEvaluation(t4[1]), ?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(t9=DagIndexerEvaluation(t0[0]), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t7=DagIndexerEvaluation(t4[1])
    TrueBranch: 10

State 10 REMAINING t6:[1..1] t1:[3..3] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] AND(?DagValueTest(t7 == ConstantValueI32(2: Int32)), t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(t9=DagIndexerEvaluation(t0[0]), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), t12=DagIndexerEvaluation(t0[-2]), ?DagValueTest(t12 == ConstantValueI32(2: Int32)), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: ?DagValueTest(t7 == ConstantValueI32(2: Int32))
    TrueBranch: 11
    FalseBranch: 14

State 11 REMAINING t6:[1..1] t1:[3..3] t7:[2..2] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] AND(t8=DagIndexerEvaluation(t4[2]), ?DagValueTest(t8 == ConstantValueI32(3: Int32)))
    2. [case [1,..[],2,3]:] AND(t9=DagIndexerEvaluation(t0[0]), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), t12=DagIndexerEvaluation(t0[-2]), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: t8=DagIndexerEvaluation(t4[2])
    TrueBranch: 12

State 12 REMAINING t6:[1..1] t1:[3..3] t7:[2..2] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] ?DagValueTest(t8 == ConstantValueI32(3: Int32))
    2. [case [1,..[],2,3]:] AND(t9=DagIndexerEvaluation(t0[0]), t10=DagSliceEvaluation(t0), t11=DagPropertyEvaluation(t10.Length), t12=DagIndexerEvaluation(t0[-2]), t13=DagIndexerEvaluation(t0[-1]), ?DagValueTest(t13 == ConstantValueI32(3: Int32)))
    Test: ?DagValueTest(t8 == ConstantValueI32(3: Int32))
    TrueBranch: 13
    FalseBranch: 14

*State 13 REMAINING t6:[1..1] t1:[3..3] t8:[3..3] t7:[2..2] t5:[3..3]
    1. [case [..[..[1,2,3]]]:] MATCH
*State 14 FAIL REMAINING t1:[3..3]

```

Reply:
>I'm not sure why case [..[..[1,2,3]]]: involves a length >= 0 check (t1=DagPropertyEvaluation(t0.Length), ?DagRelationalTest(t1 >= ConstantValueDefault(0: Int32))). Is that correct?

Sounds like we only omit tests if there's a subsuming test beforehand, not "only" because of remaining values already match.

Reply:
Turns out this affects exhaustiveness errors, so we should never emit such test.
We didn't have to deal with this situation before because we never set an "initial" remaining values.

Reply:
I updated the dump example for this test (see comment above).

I'm surprised not to see a `?DagNonNullTest(t0)` in `State 0` for both cases.

Reply:
We're using Span in this test.